README
Project Description
----------------------------------------
The Performance with Optimal Energy Toolkit (POET) provides a means to monitor
and adjust application performance while minimizing energy consumption during
application runtime.

For details, please see the following and reference as appropriate:

Connor Imes, David H. K. Kim, Martina Maggio, and Henry Hoffmann. "POET: A Portable Approach to Minimizing Energy Under Soft Real-time Constraints". In: RTAS. 2015.

Connor Imes, David H. K. Kim, Martina Maggio, and Henry Hoffmann. "Portable Multicore Resource Management for Applications with Performance Constraints". In: MCSoC. 2016.


Running POET Examples
----------------------------------------
First ensure that you have installed the Heartbeats library.

Build the library and test binaries:

$ make

Run the tests:

$ export LD_LIBRARY_PATH="./lib"
$ ./bin/[test_binary] [num_beats] [target_rate] [window_size]

The num_beats is the number of total heartbeats the application will cycle
through before exiting. The target_rate is the number of heartbeats per second,
and the window size is the number of heartbeats used to calculate the heartbeat
rate. It doesn't make sense to give the test binary a target heartbeat rate
without knowing the normal heartbeat rate of the application on your system. To
find this, run:

$ ./bin/[test_binary] [num_beats] 0 [window_size]

If you aren't sure what parameters to use, a good place to start is 200 beats
and a window size of 20.

IMPORTANT NOTE: The tests will not run until the environment variable
HEARTBEAT_ENABLED_DIR is set, which is required by the Heartbeats library.
Just set it to a temporary directory:

$ export HEARTBEAT_ENABLED_DIR=/tmp


Installing/Uninstalling POET
----------------------------------------
To install POET to the local system, run:

$ make install

Headers are installed to /usr/local/include/poet.
The library is installed to /usr/local/lib.

To uninstall POET, run:

$ make uninstall


Directory Structure
----------------------------------------
.
├── config    -- Default and example configuration files
├── inc       -- Header files
├── src       -- Source files
└── test      -- Test source files


Website
----------------------------------------
http://poet.cs.uchicago.edu/


RELEASES.md
# Release Notes

## [Unreleased]


## [v2.0.1] - 2017-10-31
### Added
 * VERSION and SOVERSION properties to shared object library
 * Multiarch support (use GNU standard installation directories)

### Changed
 * Refactor this RELEASES.md file

### Fixed
 * Fixed warnings with stricter compile flags


## [v2.0.0] - 2017-10-02
### Added
 * Added optional dependencies on heartbeats-simple and energymon in tests
 * Added Travis CI build configuration for testing

### Changed
 * Changed build system from Makefile to CMake
 * Changed log_buffer struct to poet_record struct which uses fewer memory allocations
 * Changed some label names in log files
 * Reduced logging in poet_config_linux.c:apply_cpu_config_taskset
 * Changed README to use Markdown

### Removed
 * Removed dependency on Heartbeats 2.0
 * Removed Android CMake toolchain file: cmake-toolchain/android.toolchain.cmake
 * Removed deprecated Python test wrapper

### Fixed
 * Bug fixes and optimizations


## [v1.0.1] - 2016-12-12
### Added
 * Added configuration examples for Intel Xeon E5-2690 processors
 * Added RTAS 2015 and MCSoC 2016 publication details

### Fixed
 * Fixed [#2]: Removed poet_math.h dependency in poet.h
 * Fixed [#4]: Free log buffer in poet_destroy


## 1.0 - 2015-05-29

 * Initial public release

[Unreleased]: https://github.com/libpoet/poet/compare/v2.0.1...HEAD
[v2.0.1]: https://github.com/libpoet/poet/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/libpoet/poet/compare/v1.0.1...v2.0.0
[v1.0.1]: https://github.com/libpoet/poet/compare/1.0...v1.0.1
[#4]: https://github.com/libpoet/poet/issues/4
[#2]: https://github.com/libpoet/poet/issues/2


test/math_ut.c
#include <stdio.h>
#include <stdlib.h>
#include "poet_math.h"

#define ERROR_MARGIN .0005

typedef int bool;
#define true 1
#define false 0

static double d_abs(double d) {
  return ((d > 0) ? d : -d);
}

static void assert(bool expression, char * error_message) {
  if (!expression) {
    printf("%s", error_message);
    exit(-1);
  }
}

static double fp_to_db(fp_t a) {
  return (double) (a / ((double) (1 << 16)));
}

static void addition_test(fp_t a, fp_t b, double c, double d) {
  char error_message[256];
  double expected;
  fp_t ans;
  bool expression;

  ans = FP_ADD2(a, b);
  expected = c + d;

  expression = d_abs(fp_to_db(ans) - expected) < ERROR_MARGIN;

  snprintf(error_message, sizeof(error_message), "\nExpected %f but calculated %f\n", expected, fp_to_db(ans));

  assert(expression, error_message);
}

static void subtraction_test(fp_t a, fp_t b, double c, double d) {
  char error_message[256];
  double expected;
  fp_t ans;
  bool expression;

  ans = FP_SUB(a, b);
  expected = c - d;

  expression = d_abs(fp_to_db(ans) - expected) < ERROR_MARGIN;

  snprintf(error_message, sizeof(error_message), "\nExpected %f but calculated %f\n", expected, fp_to_db(ans));

  assert(expression, error_message);
}

static void multiplication_test(fp_t a, fp_t b, double c, double d) {
  char error_message[256];
  double expected;
  fp_t ans;
  bool expression;

  ans = FP_MULT2(a, b);
  expected = DB_MULT2(c, d);

  expression = d_abs(fp_to_db(ans) - expected) < ERROR_MARGIN;

  snprintf(error_message, sizeof(error_message),
      "\nExpected %f but calculated %f with a=%f b=%f\n",
      expected, fp_to_db(ans), c, d);

  assert(expression, error_message);
}

static void division_test(fp_t a, fp_t b, double c, double d) {
  char error_message[256];
  double expected;
  fp_t ans;
  bool expression;

  ans = FP_DIV(a, b);
  expected = DB_DIV(c, d);

  expression = d_abs(fp_to_db(ans) - expected) < ERROR_MARGIN;

  snprintf(error_message, sizeof(error_message),
      "\nExpected %f but calculated %f with a=%f b=%f\n",
      expected, fp_to_db(ans), c, d);

  assert(expression, error_message);
}

static void conversion_to_fp_test(void) {
  char error_message[256];
  bool expression;
  fp_t a1;
  double a2;

  printf("Testing FP_CONST macro...");

  // Test 1
  a1 = FP_CONST(1.2);
  a2 = fp_to_db(a1);
  expression = (d_abs(1.2 - a2) < ERROR_MARGIN);
  snprintf(error_message, sizeof(error_message), "\nDouble was %f but fixed point was %f\n", 1.2, a2);
  assert(expression, error_message);

  // Test 2
  a1 = FP_CONST(89.234);
  a2 = fp_to_db(a1);
  expression = d_abs(89.234 - a2) < ERROR_MARGIN;
  snprintf(error_message, sizeof(error_message), "\nDouble was %f but fixed point was %f\n", 89.234, a2);
  assert(expression, error_message);

  // Test 3
  a1 = FP_CONST(189.24);
  a2 = fp_to_db(a1);
  expression = d_abs(189.24 - a2) < ERROR_MARGIN;
  snprintf(error_message, sizeof(error_message), "\nDouble was %f but fixed point was %f\n", 189.24, a2);
  assert(expression, error_message);

  // Test 4
  a1 = FP_CONST(-1.2);
  a2 = fp_to_db(a1);
  expression = d_abs(-1.2 - a2) < ERROR_MARGIN;
  snprintf(error_message, sizeof(error_message), "\nDouble was %f but fixed point was %f\n", -1.2, a2);
  assert(expression, error_message);

  printf("Done\n");
}

static void addition_tests(void) {
  fp_t a;
  fp_t b;

  printf("Testing fixed point addition...");

  // Test 1
  a = FP_CONST(1.52);
  b = FP_CONST(-.32);
  addition_test(a,b,1.52,-.32);

  // Test 2
  a = FP_CONST(902.314);
  b = FP_CONST(123.432);
  addition_test(a,b,902.314,123.432);

  // Test 3
  a = FP_CONST(-34.374);
  b = FP_CONST(3.432);
  addition_test(a,b,-34.374,3.432);

  // Test 4
  a = FP_CONST(-310.132);
  b = FP_CONST(-893.001);
  addition_test(a,b,-310.132,-893.001);

  // Test 5
  a = FP_CONST(20000.132);
  b = FP_CONST(20000.001);
  addition_test(a,b,32768,0);

  // Test 5
  a = FP_CONST(-20000.132);
  b = FP_CONST(-20000.001);
  addition_test(a,b,-32768,0);

  printf("Done\n");
}

static void subtraction_tests(void) {
  fp_t a;
  fp_t b;

  printf("Testing fixed point subtraction...");

  // Test 1
  a = FP_CONST(1.52);
  b = FP_CONST(-.32);
  subtraction_test(a,b,1.52,-.32);

  // Test 2
  a = FP_CONST(902.314);
  b = FP_CONST(123.432);
  subtraction_test(a,b,902.314,123.432);

  // Test 3
  a = FP_CONST(-34.374);
  b = FP_CONST(3.432);
  subtraction_test(a,b,-34.374,3.432);

  // Test 4
  a = FP_CONST(-310.132);
  b = FP_CONST(-893.001);
  subtraction_test(a,b,-310.132,-893.001);

  // Test 5
  a = FP_CONST(20000.132);
  b = FP_CONST(-20000.001);
  subtraction_test(a,b,32768,0);

  // Test 5
  a = FP_CONST(-20000.132);
  b = FP_CONST(20000.001);
  subtraction_test(a,b,-32768,0);

  printf("Done\n");
}

static void multiplication_tests(void) {
  fp_t a;
  fp_t b;

  printf("Testing fixed point multiplication...");

  // Test 1
  a = FP_CONST(1.52);
  b = FP_CONST(-.32);
  multiplication_test(a,b,1.52,-.32);

  // Test 2
  a = FP_CONST(2.314);
  b = FP_CONST(3.432);
  multiplication_test(a,b,2.314,3.432);

  // Test 3
  a = FP_CONST(-34.374);
  b = FP_CONST(3.432);
  multiplication_test(a,b,-34.374,3.432);

  // Test 4
  a = FP_CONST(-0.132);
  b = FP_CONST(-3.001);
  multiplication_test(a,b,-0.132,-3.001);

  // Test 5
  a = FP_CONST(128);
  b = FP_CONST(1024);
  multiplication_test(a,b,32768,1);

  // Test 6
  a = FP_CONST(-128);
  b = FP_CONST(-1024);
  multiplication_test(a,b,32768,1);

  // Test 7
  a = FP_CONST(-128);
  b = FP_CONST(1024);
  multiplication_test(a,b,-32768,1);

  printf("Done\n");
}

static void division_tests(void) {
  fp_t a;
  fp_t b;

  printf("Testing fixed point division...");

  // Test 1
  a = FP_CONST(1.0);
  b = FP_CONST(1.0);
  division_test(a,b,1.0,1.0);

  // Test 2
  a = FP_CONST(12.0);
  b = FP_CONST(2.0);
  division_test(a,b,12.0,2.0);

  // Test 3
  a = FP_CONST(1.5);
  b = FP_CONST(0.3);
  division_test(a,b,1.5,0.3);

  // Test 4
  a = FP_CONST(-20.0);
  b = FP_CONST(0.5);
  division_test(a,b,-20.0,0.5);

  // Test 5
  a = FP_CONST(11.34);
  b = FP_CONST(20.23);
  division_test(a,b,11.34,20.23);

  // Test 6
  a = FP_CONST(101.342);
  b = FP_CONST(002.231);
  division_test(a,b,101.342,002.231);

  // Test 7
  a = FP_CONST(-20000);
  b = FP_CONST(.001);
  division_test(a,b,-32768,1);

  // Test 8
  a = FP_CONST(20000);
  b = FP_CONST(.001);
  division_test(a,b,32768,1);

  // Test 9
  b = FP_CONST(20);
  a = FP_CONST(.001);
  division_test(a,b,1,32768);



  // Test 9
  // ints
  int32_t c = 128;
  int32_t d = 5;
  division_test(c, d, 128.0, 5.0);

  printf("Done\n");
}

int main(void) {
  printf("--------------------\nRunning Unit Tests\n--------------------\n\n");
  conversion_to_fp_test();
  addition_tests();
  subtraction_tests();
  multiplication_tests();
  division_tests();
  printf("\n");
  printf("--------------------\nFinished Unit Tests\n--------------------\n\n");

  return 0;
}


test/check_fp.py
#!/usr/bin/python

import os
import sys
import csv
import numpy

LOG_FILE = '/tmp/poet_log.txt'


class LogData:
  '''
    Simple class to read and extract data from poet and heartbeats log files
  '''

  def __init__(self):
    with open('/tmp/poet_log.txt', 'r') as fd:
      self.poet_data = self.parse(fd)

  def parse(self, fd):
    reader = csv.reader(fd, delimiter=' ', skipinitialspace=True)
    header = reader.next()
    indexes = {}
    data = {}
    for i in range(len(header)):
      indexes[i] = header[i]
      data[header[i]] = []

    for line in reader:
      for i in range(len(line)):
        header = indexes[i]
        data[header].append(float(line[i]))

    return data

  def get_windowed_hbr(self):
    return self.poet_data['HB_RATE']

  def graph(self):
    pass

if __name__ == '__main__':

  if len(sys.argv) < 2:
    print 'need file arg'
    exit()


  test = " ".join(sys.argv[1:])
  print test

  # make fixed point version
  os.system('make -s clean; make -s FIXED_POINT=1')

  # run fixed point version
  os.system(test)

  # read fixed point results results
  fixed_point_data = LogData()

  # make floating point version
  os.system('make -s clean; make -s')

  # run floating point version
  os.system(test)

  # read floating point results results
  floating_point_data = LogData()

  # assert that poet variables are close enough
  for variable in fixed_point_data.poet_data:
    x = fixed_point_data.poet_data[variable]
    y = floating_point_data.poet_data[variable]
    print 'Testing variable %s...' % (variable)
    try:
      numpy.testing.assert_array_almost_equal(x, y, decimal=2, verbose=True)
    except AssertionError as e:
      print e


test/double_loop_test.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <heartbeat-pow.h>
#include <energymon-default.h>
#include "poet.h"
#include "poet_config.h"
#include "poet_math.h"

const char* POET_LOG_FILE = "poet.log";
const unsigned int WORK_ITERATIONS = 10000000;

static inline uint64_t get_time(void) {
  struct timespec ts;
#ifdef __MACH__
  // OS X does not have clock_gettime, use clock_get_time
  clock_serv_t cclock;
  mach_timespec_t mts;
  host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
  clock_get_time(cclock, &mts);
  mach_port_deallocate(mach_task_self(), cclock);
  ts.tv_sec = mts.tv_sec;
  ts.tv_nsec = mts.tv_nsec;
#else
  // CLOCK_REALTIME is always supported, this should never fail
  clock_gettime(CLOCK_REALTIME, &ts);
#endif
  // must use a const or cast a literal - using a simple literal can overflow!
  const uint64_t ONE_BILLION = 1000000000;
  return ts.tv_sec * ONE_BILLION + ts.tv_nsec;
}

int main(int argc, char** argv) {
  if ( argc != 4 ) {
    printf("usage:\n");
    printf("processor_speed_test num_beats target_rate window_size\n");
    return -1;
  }

  uint64_t window_size = atoi(argv[3]);
  heartbeat_pow_context hb;
  heartbeat_pow_record* hb_window_buffer = malloc(window_size * sizeof(heartbeat_pow_record));
  if (hb_window_buffer == NULL) {
    perror("malloc");
    return 1;
  }
  int hb_fd = fileno(stdout);
  if (heartbeat_pow_init(&hb, window_size, hb_window_buffer, hb_fd, NULL)) {
    perror("Failed to initialize heartbeat");
    return 1;
  }
  hb_pow_log_header(hb_fd);
  energymon em;
  if (energymon_get_default(&em) || em.finit(&em)) {
    perror("Failed to initialize energymon");
    return 1;
  }

  volatile int dummy = 0;
  unsigned int num_beats = atoi(argv[1]);
  unsigned int s_nstates;
  poet_control_state_t * s_control_states;
  get_control_states("../config/default/control_config",
                     &s_control_states,
                     &s_nstates);
  poet_state * state = poet_init(CONST(atof(argv[2])),
                                 s_nstates, s_control_states, NULL,
                                 NULL,
                                 NULL,
                                 atoi(argv[3]), 1, POET_LOG_FILE);
  if (state == NULL) {
    fprintf(stderr, "Failed to initialize poet\n");
    return 1;
  }

  unsigned int i, j;
  uint64_t time_start, time_end, energy_start, energy_end;
  time_end = get_time();
  energy_end = em.fread(&em);
  for (i = 0; i < num_beats; i++) {
    time_start = time_end;
    energy_start = energy_end;
    for (j = 0; j < WORK_ITERATIONS; j++) {
      dummy = dummy >> 1;
      dummy = dummy - 1;
    }
    time_end = get_time();
    energy_end = em.fread(&em);
    heartbeat_pow(&hb, i, 1, time_start, time_end, energy_start, energy_end);
    real_t hb_window_perf = hb_pow_get_window_perf(&hb);
    real_t hb_window_power = hb_pow_get_window_power(&hb);
    poet_apply_control(state, i, hb_window_perf, hb_window_power);
  }

  poet_destroy(state);
  free(s_control_states);
  hb_pow_log_window_buffer(&hb, hb_pow_get_log_fd(&hb));
  free(hb_window_buffer);
  em.ffinish(&em);

  return 0;
}


inc/poet.h
#ifndef _POET_H
#define _POET_H

#ifdef __cplusplus
extern "C" {
#endif

#include <stdint.h>

#ifdef FIXED_POINT
typedef int32_t real_t;
#else
typedef double real_t;
#endif

/**
 * Setting this environment variable tells POET not to execute the
 * poet_apply_control function.
 * Allows disabling POET at runtime, removing the overhead of calculations.
 * Of course, no system changes will then be made either.
 */
#define POET_DISABLE_CONTROL "POET_DISABLE_CONTROL"

/**
 * Setting this environment variable tells POET not to call the apply function.
 * POET will run all its calculations but not make any system changes.
 */
#define POET_DISABLE_APPLY "POET_DISABLE_APPLY"

typedef struct poet_internal_state poet_state;

/**
 * The apply function format required to be passed to poet_init().
 */
typedef void (* poet_apply_func) (void * states,
                                  unsigned int num_states,
                                  unsigned int id,
                                  unsigned int last_id);

/**
 * The current state function is used to determine the id of the current state
 * of the system before any changes are applied. Should return -1 if the state
 * cannot be determined, 0 otherwise.
 */
typedef int (* poet_curr_state_func) (const void* states,
                                      unsigned int num_states,
                                      unsigned int* curr_state_id);

typedef struct {
  unsigned int id;
  real_t speedup;
  real_t cost;
} poet_control_state_t;

/**
 * Initializes a poet_state struct which is needed to call other functions.
 *
 * A copy of the apply_states struct will be passed to the apply function as
 * the void* parameter when it is called from poet_apply_control(). It is
 * allowed to be NULL, in which case the apply function must know where to
 * access the appropriate data structures to apply system changes.
 *
 * Default values for state variables are located in src/poet_constants.h
 *
 * @param perf_goal
 *   Must be > 0
 * @param num_system_states
 *   Must be > 0
 * @param control_states
 *   Must not be NULL
 * @param apply_states
 * @param apply
 * @param current
 * @param period
 *   Must be > 0
 * @param buffer_depth
 *   Must be > 0 if log_filename is specified
 * @param log_filename
 *
 * @return poet_state pointer, or NULL on failure (errno will be set)
 */
poet_state * poet_init(real_t perf_goal,
                       unsigned int num_system_states,
                       poet_control_state_t * control_states,
                       void * apply_states,
                       poet_apply_func apply,
                       poet_curr_state_func current,
                       unsigned int period,
                       unsigned int buffer_depth,
                       const char * log_filename);

/**
 * Deallocates memory from the poet_state struct.
 *
 * @param state
 */
void poet_destroy(poet_state * state);

/**
 * Change the performance goal at runtime.
 *
 * @param state
 * @param perf_goal
 *   Must be > 0
 */
void poet_set_performance_goal(poet_state * state,
                               real_t perf_goal);

/**
 * Runs POET decision engine and requests system changes by calling the apply
 * function provided in poet_init().
 *
 * @param state
 * @param id
 *   user-specified identifier for current iteration
 * @param perf
 *   the actual achieved performance
 * @param pwr
 *   the actual achieved power
 */
void poet_apply_control(poet_state * state,
                        unsigned long id,
                        real_t perf,
                        real_t pwr);

#ifdef __cplusplus
}
#endif

#endif


inc/poet_config.h
#ifndef _POET_CONFIG_H
#define _POET_CONFIG_H

#ifdef __cplusplus
extern "C" {
#endif

#include "poet.h"

typedef struct {
  unsigned int id;
  unsigned long freq;
  unsigned int cores;
} poet_cpu_state_t;

/**
 * Change the CPU configuration on the system by setting the frequency and
 * number of cores to be used as configured in the state with the provided id.
 *
 * Compatible with the poet_apply_func definition.
 *
 * @param states - must be a poet_cpu_state_t* (array).
 * @param num_states
 * @param id
 * @param last_id
 */
void apply_cpu_config(void* states,
                      unsigned int num_states,
                      unsigned int id,
                      unsigned int last_id);

/**
 * Read the control states from the file at the provided path and store in the
 * states pointer (states* is assigned). The number of states found is stored
 * in num_states. Returns 0 on success.
 *
 * The caller is responsible for freeing the memory this function allocates.
 *
 * @param path
 * @param states
 * @param num_states
 */
int get_control_states(const char* path,
                       poet_control_state_t** states,
                       unsigned int* num_states);

/**
 * Read the CPU states from the file at the provided path and store in the
 * states pointer (states* is assigned). The number of states found is stored
 * in num_states. Returns 0 on success.
 *
 * The caller is responsible for freeing the memory this function allocates.
 *
 * @param path
 * @param states
 * @param num_states
 */
int get_cpu_states(const char* path,
                   poet_cpu_state_t** states,
                   unsigned int* num_states);

/**
 * Attempt to determine the current system state and return the id.
 * Set curr_state_id if possible and return 0, otherwise return -1.
 *
 * Compatible with the poet_curr_state_func definition.
 *
 * @param states
 * @param num_states
 * @param curr_state_id
 */
int get_current_cpu_state(const void* states,
                          unsigned int num_states,
                          unsigned int* curr_state_id);

#ifdef __cplusplus
}
#endif

#endif


src/poet_constants.h
#ifndef _POET_CONSTANTS_H
#define _POET_CONSTANTS_H

#ifdef __cplusplus
extern "C" {
#endif

#include "poet_math.h"

static const real_t R_ZERO             =   CONST(0.0);
static const real_t R_ONE              =   CONST(1.0);
static const real_t R_TWO              =   CONST(2.0);

// estimate_base_workload constants
static const real_t X_HAT_MINUS_START  =   CONST(0.0);
static const real_t X_HAT_START        =   CONST(0.2);
static const real_t Q                  =   CONST(0.00001);
static const real_t P_START            =   CONST(1.0);
static const real_t P_MINUS_START      =   CONST(0.0);
static const real_t H_START            =   CONST(0.0);
static const real_t R                  =   CONST(0.01);
static const real_t K_START            =   CONST(0.0);

// calculate_xup constants
#define FAST 1
#define SLOW 0

#if FAST
static const real_t P1                 =   CONST(0.0);
static const real_t P2                 =   CONST(0.0);
static const real_t Z1                 =   CONST(0.0);
#elif SLOW
static const real_t P1                 =   CONST(0.1);
static const real_t P2                 =   CONST(0.8);
static const real_t Z1                 =   CONST(0.7);
#else
static const real_t P1                 =  -CONST(0.5);
static const real_t P2                 =   CONST(0.0);
static const real_t Z1                 =   CONST(0.0);
#endif

static const real_t MU                 =   CONST(1.0);
static const real_t E_START            =   CONST(1.0);
static const real_t EO_START           =   CONST(1.0);

// general constants
static const int CURRENT_ACTION_START  =  1;

#ifdef FIXED_POINT
static const real_t BIG_REAL_T         =   0x7FFFFFFF;
#else
static const real_t BIG_REAL_T         =   100000.0;
#endif

#ifdef __cplusplus
}
#endif

#endif


inc/poet_math.h
#ifndef _POET_MATH_H
#define _POET_MATH_H

#ifdef __cplusplus
extern "C" {
#endif

#include <stdio.h>
#include <stdint.h>
#include "poet.h"

/*
 * Define Q16 fixed point type
 * Define addition, subtraction, multiplication, and division operations
 * for type
 */

#define MAX_FP 0x7FFFFFFF
#define MIN_FP 0x80000000

#define UNDERFLOW_MARGIN .01

// must be aligned with FIXED_POINT real_t in poet.h, we just don't want to expose this type
typedef int32_t fp_t;

#define FP_CONST(x) ((fp_t) (((x) >= 0) ? ((x) * 65536.0 + 0.5) : ((x * 65536.0 - 0.5))))
#define DB_CONST(x) (x)

static inline fp_t FP_ADD2(fp_t a, fp_t b) {
  fp_t sum = a + b;

#ifdef POET_MATH_OVERFLOW
  // Check for overflow
  // Can only happen if a and b have the same sign,
  // and the sign of the sum is not equal to that sign
	if (!((a ^ b) & 0x80000000) && ((a ^ sum) & 0x80000000)) {
    if (a > 0) {
      sum = (fp_t) MAX_FP;
    } else {
      sum = (fp_t) MIN_FP;
    }
  }
#endif

#ifdef OVERFLOW_WARNING
  double a1 = (double) (a / ((double) (1 << 16)));
  double b1 = (double) (b / ((double) (1 << 16)));
  double answer1 = (double) (sum / ((double) (1 << 16)));

  double temp = answer1 - (a1+b1);
  if (!(temp < 1 && temp > -1)) {
    printf("\nFixed point overflow...attempted to add %f and %f\n", a1, b1);
    printf("Expected %f but received %f\n", a1+b1, answer1);
  }
#endif

  return sum;
}

static inline fp_t FP_SUB(fp_t a, fp_t b) {
  fp_t difference = a - b;

#ifdef POET_MATH_OVERFLOW
  // Check for overflow
  // Can only happen if a and b have different signs,
  // and the sign of the difference is not equal to that sign
	if (((a ^ b) & 0x80000000) && ((a ^ difference) & 0x80000000)) {
    if (a > 0) {
      difference = (fp_t) MAX_FP;
    } else {
      difference = (fp_t) MIN_FP;
    }
  }
#endif

#ifdef OVERFLOW_WARNING
  double a1 = (double) (a / ((double) (1 << 16)));
  double b1 = (double) (b / ((double) (1 << 16)));
  double answer1 = (double) (difference / ((double) (1 << 16)));

  double temp = answer1 - (a1-b1);
  if (!(temp < 1 && temp > -1)) {
    printf("\nFixed point overflow...attempted to subtract %f and %f\n", a1, b1);
    printf("Expected %f but received %f\n", a1-b1, answer1);
  }
#endif

  return difference;
}


static inline fp_t FP_MULT2(fp_t a, fp_t b) {
  int64_t prod = (int64_t) a * b;
  fp_t answer = prod >> 16;

#ifdef POET_MATH_OVERFLOW
  // Check for overflow
  // Upper 17 bits should all be the same
  // If they are not, return the maximum f
  uint32_t upper = (prod >> 47);
  if ((prod < 0 && ~upper) || (prod > 0 && upper)) {
    if ((a >= 0) == (b >= 0)) {
      answer = (fp_t) MAX_FP;
    } else {
      answer = (fp_t) MIN_FP;
    }
  }
#endif

#ifdef OVERFLOW_WARNING
  double a1 = (double) (a / ((double) (1 << 16)));
  double b1 = (double) (b / ((double) (1 << 16)));
  double answer1 = (double) (answer / ((double) (1 << 16)));

  double temp = answer1 - (a1*b1);
  if (!(temp < 1 && temp > -1)) {
    printf("\nFixed point overflow...attempted to multiply %f by %f\n", a1, b1);
    printf("Expected %f but received %f\n", a1*b1, answer1);
    return answer;
  }
#endif

#ifdef UNDERFLOW_WARNING
  if (a == 0 || b == 0) {
    return answer;
  }


  double a2 = (double) (a / ((double) (1 << 16)));
  double b2 = (double) (b / ((double) (1 << 16)));
  double answer2 = (double) (answer / ((double) (1 << 16)));

  double temp2 = 1 - answer2 / (a2*b2);
  if (!(temp2 < UNDERFLOW_MARGIN && temp > -UNDERFLOW_MARGIN)) {
    printf("\nFixed point underflow...attempted to multiply %f by %f\n", a2, b2);
    printf("Expected %f but received %f\n", a2*b2, answer2);
  }
#endif

  return answer;
}

static inline fp_t FP_DIV(fp_t a, fp_t b) {
  int64_t quotient = ((((int64_t) a) << 16) / b);
  fp_t answer = (fp_t) quotient;

#ifdef POET_MATH_OVERFLOW
  // Check for overflow
  // Upper 17 bits should all be the same
  // If they are not, return the maximum f
  uint32_t upper = (quotient >> 31);
  if ((answer < 0 && ~upper) || (answer > 0 && upper)) {
    if ((a >= 0) == (b >= 0)) {
      answer = (fp_t) MAX_FP;
    } else {
      answer = (fp_t) MIN_FP;
    }
  }
#endif

#ifdef OVERFLOW_WARNING
  double a1 = (double) (a / ((double) (1 << 16)));
  double b1 = (double) (b / ((double) (1 << 16)));
  double answer1 = (double) (answer / ((double) (1 << 16)));

  double temp = answer1 - (a1/b1);
  if (!(temp < 1 && temp > -1)) {
    printf("\nFixed point overflow...attempted to divide %f by %f\n", a1, b1);
    printf("Expected %f but received %f\n", a1/b1, answer1);
    return answer;
  }
#endif

#ifdef UNDERFLOW_WARNING
  if (a == 0) {
    return answer;
  }

  double a2 = (double) (a / ((double) (1 << 16)));
  double b2 = (double) (b / ((double) (1 << 16)));
  double answer2 = (double) (answer / ((double) (1 << 16)));

  double temp2 = 1 - answer2 / (a2/b2);
  if (!(temp2 < UNDERFLOW_MARGIN && temp > -UNDERFLOW_MARGIN)) {
    printf("\nFixed point underflow...attempted to divide %f by %f\n", a2, b2);
    printf("Expected %f but received %f\n", a2/b2, answer2);
  }
#endif

  return answer;
}

// For ease of use, multiply >2 fixed point values at once
#define FP_MULT3(a, b, c) (FP_MULT2(FP_MULT2((a), (b)), (c)))
#define FP_MULT4(a, b, c, d) (FP_MULT2(FP_MULT3((a), (b), (c)), (d)))

// Same operations for floating point numbers
#define DB_MULT2(a, b) ((a) * (b))
#define DB_MULT3(a, b, c) (DB_MULT2(DB_MULT2((a), (b)), (c)))
#define DB_MULT4(a, b, c, d) (DB_MULT2(DB_MULT3((a), (b), (c)), (d)))
#define DB_DIV(a, b) ((a) / (b))

/*
 * If fixed point is defined, use fixed point functions
 * Otherwise use normal floating point functions
 */

#ifdef FIXED_POINT

#define CONST(x) FP_CONST((x))
#define mult(a,b) (FP_MULT2((a), (b)))
#define mult3(a,b,c) (FP_MULT3((a), (b), (c)))
#define mult4(a,b,c,d) (FP_MULT4((a), (b), (c), (d)))
#define div(a,b) (FP_DIV((a),(b)))

#define int_to_real(a) ((real_t) (a) << 16)
#define real_to_db(a) ((double) ((a) / ((double) (1 << 16))))
#define real_to_int(a) (((a) + CONST(.5)) >> 16)

#else

#define CONST(x) DB_CONST((x))
#define mult(a,b) (DB_MULT2((a), (b)))
#define mult3(a,b,c) (DB_MULT3((a), (b), (c)))
#define mult4(a,b,c,d) (DB_MULT4((a), (b), (c), (d)))
#define div(a,b) (DB_DIV((a),(b)))

#define int_to_real(a) ((double) (a))
#define real_to_db(a) (a)
#define real_to_int(a) ((a) + .5)

#endif

#ifdef __cplusplus
}
#endif

#endif
